### PR TITLE
ffi: Move `SchemaTreeNode` into `SchemaTree`; Use an `optional` for the parent ID in the schema tree node to track nodes which have no parents (the root).

### PR DESCRIPTION
**Pull request by LinZhihao-723**
# Description

In the current implementation, each ffi schema tree node stores an ID to keep track of the parent node ID. However, this scheme doesn't make sense for the root node since the root, by definition, should have no parent. In the current implementation, we manually set the parent ID of the root to itself (which is 0).
This PR changes the underlying parent ID from `id_t` to `std::optional<id_t>` so the parent ID can be null. This scheme allows us to implement the root properly by setting its parent ID to null. After this change, there are two standard approaches to identifying whether a node is a root:
1. If the node has no parent ID (null), it is identified as the root.
2. If the node's ID matches the root ID defined by `SchemaTree`, it is identified as the root.

We then made the following changes according to these two root identifications:
- If the code refers to a schema tree node instance, i.e., when traversing the schema tree, we use the first approach.
- If the code refers to the node by a numerical ID,i.e., constructing a node locator, we use the second approach.

Notice that either of these approaches is valid if and only if the root is unique in a schema tree. It is the schema tree's responsibility to ensure the root's validity and uniqueness. We secured this with the following design:
- `SchemaTreeNode` is instantiated using factory functions. There are two factory functions: one for the root node and one for non-root nodes. The constructed node should be either a root node or a non-root node.
- `SchemaTree` calls the root node factory functions to create the root as its first node (ID 0) on initialization. The root factory function must not be called elsewhere.
- `SchemaTree` inserts new nodes to the tree identified by a valid node locator. Since the node locator specifies a parent ID, conceptually, the node created by such a locator must be a non-root node.

To achieve this design, we move `SchemaTreeNode` into the schema tree class as `SchemaTree::Node`, so that it can access `SchemaTree::cRootId` and `SchemaTree::NodeLocator` (to implement bullet point 2 and 3 above) without circular dependency. The factory functions and constructors of the node class are made private. `SchemaTree` is marked as `friend` to ensure it is the only class that can instantiate tree nodes.

We update the schema tree unit test to ensure that each build meets the assumptions above in case future development on the schema tree breaks them.

# Validation performed

- Ensure workflows passed
- Ensure existing unit tests aren't broken
- Ensure newly added unit tests passed

## Summary by CodeRabbit

- **New Features**
	- Enhanced encapsulation and clarity in schema tree node management.
	- Introduced a new `Node` class to improve the structure of schema tree nodes.
	- Added profiling options and configuration settings for improved build management.
	- Default build type set to "Release" for better user experience.

- **Bug Fixes**
	- Improved error handling for unsupported compressor values and missing libraries.

- **Tests**
	- Updated tests to reflect changes in node identifiers and root node management.

### CODE REVIEW

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp [hunk 1]
```diff
@@ -294,17 +297,21 @@ auto get_schema_subtree_bitmap(
...
-        auto parent_id{schema_tree.get_node(node_id).get_parent_id()};
+        auto optional_parent_id{schema_tree.get_node(node_id).get_parent_id()};
         while (true) {
+            if (false == optional_parent_id.has_value()) {
+                // Root has been reached.
+                // Ideally this if statement can be used as the loop condition. However, clang-tidy
+                // will complain about unchecked optional access.
+                break;
+            }
+            auto const parent_id{optional_parent_id.value()};
```
[the remark this answers is not in the document]
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:307]: as explained in the comment, this is to silent clang-tidy

File: components/core/src/clp/ffi/SchemaTreeNode.hpp [hunk 2]
```diff
@@ -77,8 +112,15 @@
     }
 
 private:
+    // Constructors
+    SchemaTreeNode(id_t id, std::optional<id_t> parent_id, std::string_view key_name, Type type)
+            : m_id{id},
+              m_parent_id{parent_id},
+              m_key_name{key_name.begin(), key_name.end()},
+              m_type{type} {}
+
```
[the remark this answers is not in the document]
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/SchemaTreeNode.hpp]: at this moment we don't consider to derive this class.

File: components/core/src/clp/ffi/SchemaTreeNode.hpp [hunk 3]
```diff
@@ -49,7 +67,24 @@
     // Methods
     [[nodiscard]] auto get_id() const -> id_t { return m_id; }
 
-    [[nodiscard]] auto get_parent_id() const -> id_t { return m_parent_id; }
+    [[nodiscard]] auto is_root() const -> bool { return false == m_parent_id.has_value(); }
+
```
[the remark this answers is not in the document]
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/SchemaTreeNode.hpp]: yoda condition is our coding standard.

File: components/core/src/clp/ffi/SchemaTreeNode.hpp [hunk 4]
```diff
@@ -49,7 +67,24 @@
...
+    /**
+     * Gets the parent ID without checking if it is std::nullopt.
+     * NOTE: This method should only be used if the caller has checked the node is not the root.
+     * `is_root()` must return false to call this method safely.
+     * @return the ID of the parent node in the schema tree.
+     */
+    [[nodiscard]] auto get_parent_id_unsafe() const -> id_t {
+        // NOLINTNEXTLINE(bugprone-unchecked-optional-access)
+        return m_parent_id.value();
+    }
```
[the remark this answers is not in the document]
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/SchemaTreeNode.hpp]: as suggested it is unsafe. We don't add assertions in our code.

File: components/core/src/clp/ffi/SchemaTreeNode.hpp [hunk 5]
```diff
@@ -28,12 +29,29 @@ class SchemaTreeNode {
...
+    /**
+     * Creates a root node.
+     * @param id
+     * @param parent_id
+     * @param key_name
+     * @param type
+     */
+    [[nodiscard]] static auto create_root(id_t id) -> SchemaTreeNode {
+        return {id, std::nullopt, "", Type::Obj};
+    }
```
[the remark this answers is not in the document]
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/SchemaTreeNode.hpp]: Thx. This should be updated in the latest commit.

File: components/core/src/clp/ffi/SchemaTree.cpp [hunk 6]
```diff
@@ -43,12 +43,12 @@ auto SchemaTree::insert_node(NodeLocator const& locator) -> SchemaTreeNode::id_t
...
     }
     auto const node_id{static_cast<SchemaTreeNode::id_t>(m_tree_nodes.size())};
-    m_tree_nodes.emplace_back(
+    m_tree_nodes.emplace_back(SchemaTreeNode::create(
             node_id,
             locator.get_parent_id(),
             locator.get_key_name(),
             locator.get_type()
-    );
+    ));
```
[the remark this answers is not in the document]
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/SchemaTree.cpp]: locator's ID is not `std::optional`

File: components/core/src/clp/ffi/ir_stream/Serializer.cpp [hunk 7]
```diff
@@ -292,7 +292,7 @@ auto Serializer<encoded_variable_t>::serialize_msgpack_map(msgpack::object_map c
     bool failure{false};
     vector<MsgpackMapIterator> dfs_stack;
     dfs_stack.emplace_back(
-            SchemaTree::cRootId,
+            SchemaTree::get_root_node_id(),
```
[the remark this answers is not in the document]
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/ir_stream/Serializer.cpp:294]: it's in the header of `SchemaTree`

File: components/core/src/clp/ffi/SchemaTree.cpp [hunk 8]
```diff
@@ -38,19 +36,14 @@ auto SchemaTree::try_get_node_id(NodeLocator const& locator
...
-            locator.get_parent_id(),
-            locator.get_key_name(),
-            locator.get_type()
-    );
+    auto const node_id{static_cast<Node::id_t>(m_tree_nodes.size())};
+    m_tree_nodes.emplace_back(Node::create(node_id, locator));
     auto& parent_node{m_tree_nodes[locator.get_parent_id()]};
-    if (SchemaTreeNode::Type::Obj != parent_node.get_type()) {
+    if (Node::Type::Obj != parent_node.get_type()) {
         throw OperationFailed(
```
[the remark this answers is not in the document]
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/SchemaTree.cpp:47]: The parent id of locator is not `std::optional`

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp [hunk 9]
```diff
@@ -49,7 +48,7 @@ class JsonSerializationIterator {
 public:
     // Constructor
     JsonSerializationIterator(
-            SchemaTreeNode const* schema_tree_node,
+            SchemaTree::Node const* schema_tree_node,
```
[the remark this answers is not in the document]
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:51]: not necessary in this use case.

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp [hunk 10]
```diff
@@ -100,16 +99,16 @@
      * Gets the next child schema tree node and advances the iterator.
      * @return The next child schema tree node.
      */
-    [[nodiscard]] auto get_next_child_schema_tree_node() -> SchemaTreeNode::id_t {
+    [[nodiscard]] auto get_next_child_schema_tree_node() -> SchemaTree::Node::id_t {
```
[the remark this answers is not in the document]
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:102]: The code calls `has_next_child_schema_tree_node` before calling `get_next_child_schema_tree_node`

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp [hunk 11]
```diff
@@ -100,16 +99,16 @@
...
 
     [[nodiscard]] auto get_json_obj() -> nlohmann::json::object_t& { return m_json_obj; }
 
 private:
-    SchemaTreeNode const* m_schema_tree_node;
-    vector<SchemaTreeNode::id_t> m_child_schema_tree_nodes;
-    vector<SchemaTreeNode::id_t>::const_iterator m_child_schema_tree_node_it;
+    SchemaTree::Node const* m_schema_tree_node;
+    vector<SchemaTree::Node::id_t> m_child_schema_tree_nodes;
+    vector<SchemaTree::Node::id_t>::const_iterator m_child_schema_tree_node_it;
```
[the remark this answers is not in the document]
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:111]: no need in this use case

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp [hunk 12]
```diff
@@ -215,33 +214,36 @@ auto validate_node_id_value_pairs(
...
 ) -> std::errc {
     try {
-        std::unordered_map<SchemaTreeNode::id_t, std::unordered_set<std::string_view>>
+        std::unordered_map<SchemaTree::Node::id_t, std::unordered_set<std::string_view>>
                 parent_node_id_to_key_names;
         for (auto const& [node_id, value] : node_id_value_pairs) {
-            if (SchemaTree::cRootId == node_id) {
+            auto const& node{schema_tree.get_node(node_id)};
+            if (node.is_root()) {
+                // This node is the root
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:222]:
```suggestion
```

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp [hunk 13]
```diff
@@ -215,33 +214,36 @@ auto validate_node_id_value_pairs(
...
+            if (SchemaTree::Node::Type::Obj == node_type
                 && false == is_leaf_node(schema_tree, node_id, node_id_value_pairs))
             {
                 // The node's value is `null` or `{}` but its descendants appear in
                 // `node_id_value_pairs`.
                 return std::errc::operation_not_permitted;
             }
 
-            auto const parent_node_id{node.get_parent_id()};
+            // Since we checked that the node must not be the root, we can query the underlying ID
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:244]:
```suggestion
            // We checked that the node isn't the root above, so we can query the underlying ID
```

File: components/core/src/clp/ffi/KeyValuePairLogEvent.cpp [hunk 14]
```diff
@@ -294,24 +296,28 @@ auto get_schema_subtree_bitmap(
         schema_subtree_bitmap[node_id] = true;
 
         // Iteratively mark the parents as true
-        auto parent_id{schema_tree.get_node(node_id).get_parent_id()};
+        auto optional_parent_id{schema_tree.get_node(node_id).get_parent_id()};
         while (true) {
+            if (false == optional_parent_id.has_value()) {
+                // Root has been reached.
+                // Ideally this if statement can be used as the loop condition. However, clang-tidy
+                // will complain about unchecked optional access.
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/KeyValuePairLogEvent.cpp:304]:
```suggestion
            // Ideally, we'd use this if statement as the loop condition, but clang-tidy will
            // complain about an unchecked `optional` access.
            if (false == optional_parent_id.has_value()) {
                // Reached the root
```

File: components/core/src/clp/ffi/SchemaTree.hpp [hunk 15]
```diff
@@ -91,39 +91,153 @@ class SchemaTree {
...
+        ~Node() = default;
+
+        // Methods
+        [[nodiscard]] auto get_id() const -> id_t { return m_id; }
+
+        [[nodiscard]] auto is_root() const -> bool { return false == m_parent_id.has_value(); }
+
+        /**
+         * @return the ID of the parent node in the schema tree if the node is not root.
+         * @return std::nullopt if the node is root.
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/SchemaTree.hpp:137]:
```suggestion
         * @return The ID of the parent node in the schema tree, if the node is not the root.
         * @return std::nullopt if the node is the root.
```

File: components/core/src/clp/ffi/SchemaTree.hpp [hunk 16]
```diff
@@ -91,39 +91,153 @@ class SchemaTree {
...
+         */
+        [[nodiscard]] auto get_parent_id() const -> std::optional<id_t> { return m_parent_id; }
+
+        /**
+         * Gets the parent ID without checking if it is std::nullopt.
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/SchemaTree.hpp:142]:
```suggestion
         * Gets the parent ID without checking if it's `std::nullopt`.
```

File: components/core/src/clp/ffi/SchemaTree.hpp [hunk 17]
```diff
@@ -91,39 +91,153 @@ class SchemaTree {
...
+         * NOTE: This method should only be used if the caller has checked the node is not the root.
+         * `is_root()` must return false to call this method safely.
+         * @return the ID of the parent node in the schema tree.
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/SchemaTree.hpp:145]:
```suggestion
         * @return The ID of the parent node in the schema tree.
```

File: components/core/src/clp/ffi/SchemaTree.hpp [hunk 18]
```diff
@@ -91,39 +91,153 @@ class SchemaTree {
...
+         * @param id
+         * @param locator
+         */
+        [[nodiscard]] static auto create(id_t id, NodeLocator const& locator) -> Node {
+            return {id, locator.get_parent_id(), locator.get_key_name(), locator.get_type()};
+        }
+
+        /**
+         * Creates a root node.
+         * @param id The ID of the root node assigned by the schema tree.
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/SchemaTree.hpp:192]:
```suggestion
```

File: components/core/src/clp/ffi/SchemaTree.hpp [hunk 19]
```diff
@@ -91,39 +91,153 @@ class SchemaTree {
...
+         */
+        [[nodiscard]] static auto create_root() -> Node {
+            return {cRootId, std::nullopt, "", Type::Obj};
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/SchemaTree.hpp:195]:
```suggestion
            return {cRootId, std::nullopt, {}, Type::Obj};
```

File: components/core/src/clp/ffi/SchemaTree.hpp [hunk 20]
```diff
@@ -91,39 +91,153 @@ class SchemaTree {
...
+        Node::id_t m_parent_id;
         std::string_view m_key_name;
-        SchemaTreeNode::Type m_type;
+        Node::Type m_type;
     };
 
-    // Constants
-    static constexpr SchemaTreeNode::id_t cRootId{0};
+    // Static methods
+    [[nodiscard]] constexpr static auto get_root_node_id() -> Node::id_t { return cRootId; }
```
Comment (kirkrodrigues) [inline components/core/src/clp/ffi/SchemaTree.hpp:237]: Why not leave `cRootId` public?

File: components/core/tests/test-ffi_SchemaTree.cpp [hunk 21]
```diff
@@ -19,41 +17,63 @@ namespace {
...
+        SchemaTree::Node::id_t expected_id
 ) -> bool;
 
 /**
  * @param schema_tree
  * @param locator
  * @param expected_id
- * @return Whether the node exists and its ID matches the expected ID.
+ * @return Whether the node exists with its ID matches the expected ID and its information matches
+ * the ones specified in the locator.
```
Comment (kirkrodrigues) [inline components/core/tests/test-ffi_SchemaTree.cpp:29]:
```suggestion
 * @return Whether an ID could be found for a node matching the locator, the ID matches the expected
 * ID, the corresponding node is not the root, and it matches the locator.
```

File: components/core/tests/test-ffi_SchemaTree.cpp [hunk 22]
```diff
@@ -19,41 +17,63 @@ namespace {
...
  */
 [[nodiscard]] auto check_node(
```
Comment (kirkrodrigues) [inline components/core/tests/test-ffi_SchemaTree.cpp:31]: How about `check_non_root_node`?

File: components/core/tests/test-ffi_SchemaTree.cpp [hunk 23]
```diff
@@ -19,41 +17,63 @@ namespace {
...
         SchemaTree const& schema_tree,
-        SchemaTree::NodeLocator locator,
-        SchemaTreeNode::id_t expected_id
+        SchemaTree::NodeLocator const& locator,
+        SchemaTree::Node::id_t expected_id
 ) -> bool {
     auto const node_id{schema_tree.try_get_node_id(locator)};
-    return node_id.has_value() && node_id.value() == expected_id;
+    if (false == node_id.has_value() || node_id.value() != expected_id) {
+        // Check node ID
```
Comment (kirkrodrigues) [inline components/core/tests/test-ffi_SchemaTree.cpp:53]:
```suggestion
        // The node's ID doesn't match.
```

File: components/core/tests/test-ffi_SchemaTree.cpp [hunk 24]
```diff
@@ -19,41 +17,63 @@ namespace {
...
+        return false;
+    }
+    auto const& node{schema_tree.get_node(expected_id)};
+    if (node.is_root()) {
+        // The node after initialization must not be the root.
```
Comment (kirkrodrigues) [inline components/core/tests/test-ffi_SchemaTree.cpp:58]:
```suggestion
        // Any nodes added after the tree was constructed must not be the root.
```

File: components/core/tests/test-ffi_SchemaTree.cpp [hunk 25]
```diff
@@ -19,41 +17,63 @@ namespace {
...
+        return false;
+    }
+    auto const optional_parent_id{node.get_parent_id()};
+    if (false == optional_parent_id.has_value()) {
+        // Non-root node must have a parent ID.
```
Comment (kirkrodrigues) [inline components/core/tests/test-ffi_SchemaTree.cpp:63]:
```suggestion
        // Non-root nodes must have a parent ID.
```

File: components/core/tests/test-ffi_SchemaTree.cpp [hunk 26]
```diff
@@ -19,41 +17,63 @@ namespace {
...
+        return false;
+    }
+    if (optional_parent_id.value() != locator.get_parent_id()
+        || node.get_type() != locator.get_type() || node.get_key_name() != locator.get_key_name())
```
Comment (kirkrodrigues) [inline components/core/tests/test-ffi_SchemaTree.cpp:67]: Technically this is implicitly checked by `try_get_node`.

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

File: components/core/tests/test-ffi_SchemaTree.cpp (hunk 26 above)
Comment (LinZhihao-723) [reply components/core/tests/test-ffi_SchemaTree.cpp:67]: that is correct, but adding this in the unit test is also checking the functionality of `try_get_node` works properly right?

File: components/core/src/clp/ffi/SchemaTree.hpp (hunk 20 above)
Comment (LinZhihao-723) [reply components/core/src/clp/ffi/SchemaTree.hpp]: discussed offline and realized using public `cRootId` vs. public method makes no difference.

**Review by kirkrodrigues — APPROVED** [review]
For the PR title, how about:

```
ffi: Move `SchemaTreeNode` into `SchemaTree`; Use an `optional` for the parent ID in the schema tree node to track nodes which have no parents (the root).
```

**Pull request merged**